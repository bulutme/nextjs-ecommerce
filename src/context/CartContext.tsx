import React, {
  createContext,
  useContext,
  ReactNode,
  useEffect,
  useState,
} from "react";
import { ProductItemProps } from "@/app/types/Product/types";
import useLocalStorage from "@/hooks/useLocalStorage";

type CartItem = {
  product: ProductItemProps;
  count: number;
};

type GroupedCart = {
  brand: string;
  products: CartItem[];
};

type CartContextType = {
  cart: GroupedCart[];
  addToCart: (product: ProductItemProps) => void;
  removeProduct: (productId: number) => void;
  increaseCount: (productId: number) => void;
  decreaseCount: (productId: number) => void;
  totalItemCount: number;
  totalPrice: string;
  isLoading: boolean;
};

const CartContext = createContext<CartContextType | undefined>(undefined);

const CartProvider = ({ children }: { children: ReactNode }) => {
  const [cart, setCart] = useLocalStorage<GroupedCart[]>("cart", []);
  const [isLoading, setIsLoading] = useState<boolean>(true);

  useEffect(() => {
    if (typeof window !== "undefined") {
      setIsLoading(true);
      setCart(JSON.parse(localStorage.getItem("cart") as string));
      setIsLoading(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const totalItemCount = cart.reduce(
    (total, group) =>
      total +
      group.products.reduce((groupTotal, item) => groupTotal + item.count, 0),
    0
  );

  const totalPrice = cart.reduce(
    (total, group) =>
      total +
      group.products.reduce(
        (groupTotal, item) => groupTotal + item.count * item.product.price,
        0
      ),
    0
  );

  const formattedTotalPrice = totalPrice.toFixed(2);

  const addToCart = (product: ProductItemProps) => {
    const existingGroupIndex = cart.findIndex(
      (group) => group.brand === product.brand
    );

    if (existingGroupIndex !== -1) {
      const existingProductIndex = cart[existingGroupIndex].products.findIndex(
        (item) => item.product.id === product.id
      );

      const newCart = [...cart];

      if (existingProductIndex !== -1) {
        const cartId =
          newCart[existingGroupIndex].products[existingProductIndex].product.id;
        increaseCount(cartId);
      } else {
        newCart[existingGroupIndex].products.push({
          product,
          count: 1,
        });
        setCart(newCart);
      }
    } else {
      setCart((prevCart) => [
        ...prevCart,
        {
          brand: product.brand,
          products: [
            {
              product,
              count: 1,
            },
          ],
        },
      ]);
    }
  };

  const removeProduct = (productId: number) => {
    const newCart = cart.map((group) => ({
      ...group,
      products: group.products.filter((item) => item.product.id !== productId),
    }));

    // Remove empty groups
    const updatedCart = newCart.filter((group) => group.products.length > 0);
    setCart(updatedCart);
  };

  const increaseCount = (productId: number) => {
    const newCart = cart.map((group) => {
      return {
        ...group,
        products: group.products.map((item) =>
          item.product.id === productId
            ? { ...item, count: item.count + 1 }
            : item
        ),
      };
    });

    setCart(newCart);
  };

  const decreaseCount = (productId: number) => {
    const newCart = cart.map((group) => ({
      ...group,
      products: group.products.map((item) =>
        item.product.id === productId
          ? { ...item, count: Math.max(1, item.count - 1) }
          : item
      ),
    }));

    setCart(newCart);
  };

  const contextValue: CartContextType = {
    cart,
    totalItemCount,
    totalPrice: formattedTotalPrice,
    addToCart,
    removeProduct,
    increaseCount,
    decreaseCount,
    isLoading,
  };

  return (
    <CartContext.Provider value={contextValue}>{children}</CartContext.Provider>
  );
};

const useCart = () => {
  const context = useContext(CartContext);

  if (!context) {
    throw new Error("useCart must be used within a CartProvider");
  }
  return context;
};

export { CartProvider, useCart };
