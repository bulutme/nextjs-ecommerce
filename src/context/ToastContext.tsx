import { createContext, FC, ReactNode, useContext, useState } from "react";

export type ToastType = "success" | "error" | "exit";

type ToastData = {
  id: number;
  type: ToastType;
  message: string;
};

type ToastContextType = {
  toasts: ToastData[];
  addToast: (type: ToastType, message: string) => void;
  removeToast: (id: number) => void;
};

const MAX_TOASTS = 5;

const ToastContext = createContext<ToastContextType | undefined>(undefined);

export const ToastProvider: FC<{ children: ReactNode }> = ({ children }) => {
  const [toasts, setToasts] = useState<ToastData[]>([]);

  // function to add toast using type and message
  const addToast = (type: ToastType, message: string) => {
    const id = new Date().getTime();
    const newToast: ToastData = { id, type, message };

    // remove the latest one if toast item more than 5
    setToasts((prevToasts) => [...prevToasts.slice(-MAX_TOASTS + 1), newToast]);

    if (type !== "exit") {
      setTimeout(() => {
        removeToast(id);
      }, 3000);
    }
  };

  // function to remove toast
  const removeToast = (id: number) => {
    setToasts((prevToasts) => prevToasts.filter((toast) => toast.id !== id));
  };

  const contextValue: ToastContextType = {
    toasts,
    addToast,
    removeToast,
  };

  return (
    <ToastContext.Provider value={contextValue}>
      {children}
    </ToastContext.Provider>
  );
};

// custom hook to use toast context
export const useToast = () => {
  const context = useContext(ToastContext);
  if (!context) {
    throw new Error("useToast must be used within a ToastProvider");
  }
  return context;
};
