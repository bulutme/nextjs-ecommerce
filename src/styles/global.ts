"use client";

import { createGlobalStyle } from "styled-components";

const GlobalStyles = createGlobalStyle`
  * {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
  }

  html {
    font-size: 16px;
  }

  html, body, #__next {
    height: 100%;
  }

  a {
    text-decoration: none;
    width: fit-content;
  }

  ul {
    list-style-type: none
  }
`;

export default GlobalStyles;
