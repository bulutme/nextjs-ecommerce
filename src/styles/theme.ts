export const theme = {
  colors: {
    black: "#000",
    white: "#fff",
    gray: "#808080",
    grayLight: "#d3d3d3",
    border: "#e2e8f0",
    primary: "#3bb77e",
    primaryDark: "#2b8a5e",
    error: "#dc2626",
    info: "#0891b2",
    success: "#16a34a",
  },
  screens: {
    "2xs": "400px",
    xs: "480px",
    sm: "576px",
    md: "768px",
    lg: "992px",
    xl: "1200px",
    "2xl": "1440px",
    "3xl": "1560px",
  },
  fontWeight: {
    normal: 400,
    medium: 500,
    semibold: 600,
    bold: 700,
  },
};
