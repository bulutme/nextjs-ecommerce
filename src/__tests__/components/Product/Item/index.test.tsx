import { render, screen, fireEvent } from "@testing-library/react";
import { CartProvider } from "@/context/CartContext";
import { ToastProvider } from "@/context/ToastContext";
import ProductItem from "@/components/Product/Item";
import { ThemeProvider } from "styled-components";
import { theme } from "@/styles/theme";

jest.mock("../../../../hooks/useLocalStorage", () => ({
  __esModule: true,
  default: jest.requireActual("../../../../hooks/useLocalStorage").default,
  useLocalStorage: jest.requireActual("../../../../hooks/useLocalStorage")
    .useLocalStorage,
}));

const sampleProduct = {
  id: 1,
  category: "Electronics",
  brand: "Sample Brand",
  price: 29.99,
  image: "https://source.unsplash.com/random/300Ã—300",
  description: "Sample description",
  name: "Sample Product",
};

test("renders product information and add to cart button", () => {
  render(
    <ThemeProvider theme={theme}>
      <CartProvider>
        <ToastProvider>
          <ProductItem {...sampleProduct} />
        </ToastProvider>
      </CartProvider>
    </ThemeProvider>
  );

  const brandElement = screen.getByText(sampleProduct.brand);
  const nameElement = screen.getByText(sampleProduct.name);
  const categoryElement = screen.getByText(sampleProduct.category);
  const priceElement = screen.getByText(`$${sampleProduct.price}`);
  const addToCartButton = screen.getByText(/add to cart/i);

  expect(brandElement).toBeInTheDocument();
  expect(nameElement).toBeInTheDocument();
  expect(categoryElement).toBeInTheDocument();
  expect(priceElement).toBeInTheDocument();
  expect(addToCartButton).toBeInTheDocument();
});

test("matches snapshot when Add to Cart button is clicked", () => {
  const { container } = render(
    <ThemeProvider theme={theme}>
      <CartProvider>
        <ToastProvider>
          <ProductItem {...sampleProduct} />
        </ToastProvider>
      </CartProvider>
    </ThemeProvider>
  );

  const addToCartButton = screen.getByText(/add to cart/i);

  fireEvent.click(addToCartButton);

  expect(container).toMatchSnapshot();
});
