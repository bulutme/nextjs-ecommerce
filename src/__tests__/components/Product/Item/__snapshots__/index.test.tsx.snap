// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot when Add to Cart button is clicked 1`] = `
<div>
  <div
    class="sc-6a90263a-1 kPhoNf"
  >
    <div
      class="sc-6a90263a-2 gmckNz"
    >
      <img
        alt="Sample Product"
        class="sc-6a90263a-3 blhajF"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height="300"
        src="/_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Frandom%2F300%C3%97300&w=1080&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Frandom%2F300%C3%97300&w=640&q=75 1x, /_next/image?url=https%3A%2F%2Fsource.unsplash.com%2Frandom%2F300%C3%97300&w=1080&q=75 2x"
        style="color: transparent;"
        width="500"
      />
      <p
        class="sc-6a90263a-0 ZZjYn"
      >
        Sample description
      </p>
    </div>
    <div
      class="sc-6a90263a-4 eddXMR"
    >
      <p
        class="sc-6a90263a-6 edlzUM"
      >
        Sample Brand
      </p>
      <h3
        class="sc-6a90263a-5 hqOZZq"
      >
        Sample Product
      </h3>
      <p
        class="sc-6a90263a-7 bUwdxL"
      >
        Electronics
      </p>
      <div
        class="sc-6a90263a-8 lnAPoO"
      >
        <span
          class="sc-6a90263a-9 iKWciJ"
        >
          $
          29.99
        </span>
        <button
          class="sc-e6c76d24-0 iuIUPx"
          color="primary"
        >
          Add to cart
        </button>
      </div>
    </div>
  </div>
</div>
`;
